src/java.net.http/share/classes/jdk/internal/net/http/PlainHttpConnection.java

```diff
@@ -187,7 +201,43 @@ public void abort(IOException ioe) {
                     debug.log("Failed to close channel after unsuccessful connect");
             }
         }
-        return cf;
+        return cf.handle((r,t) -> checkRetryConnect(r, t,exchange))
+                .thenCompose(Function.identity());
```

Comment:
So this changes behaviour so that a single subsequent additional TCP connection may be tried for all cases where the connection fails AND there is remaining connection deadline. Ok.

Reply:
Yes - I didn't want to depend on the exception message - and there's no way to distinguish with the exception type (it's raw `ConnectionException`). It will however happen *only* when the `ConnectionException` is thrown by `SocketChannel::finishConnect`.

Reply:
@ChrisHegarty I have modified the logic to not retry if `jdk.httpclient.disableRetryConnect` was specified. By default that property is not specified and so we will retry once, but if an application has specifically opted for not retrying on ConnectException then we will not retry.